core/lib/state/src/storage_factory.rs

```diff
@@ -65,19 +69,34 @@ pub struct RocksdbWithMemory {
     pub batch_diffs: Vec<BatchDiff>,
 }
 
-/// A [`ReadStorage`] implementation that uses either [`PostgresStorage`] or [`RocksdbStorage`]
-/// underneath.
+/// Union of all [`ReadStorage`] implementations that are returned by [`ReadStorageFactory`], such as
+/// Postgres- and RocksDB-backed storages.
+///
+/// Ordinarily, you might want to use the [`OwnedStorage`] type alias instead of using `CommonStorage` directly.
+/// The former naming signals that the storage has static lifetime and thus can be sent to Tokio tasks or other threads.
 #[derive(Debug)]
-pub enum PgOrRocksdbStorage<'a> {
+pub enum CommonStorage<'a> {
     /// Implementation over a Postgres connection.
     Postgres(PostgresStorage<'a>),
     /// Implementation over a RocksDB cache instance.
     Rocksdb(RocksdbStorage),
     /// Implementation over a RocksDB cache instance with in-memory DB diffs.
     RocksdbWithMemory(RocksdbWithMemory),
+    /// In-memory storage snapshot with the Postgres storage fallback.
+    Snapshot(StorageWithSnapshot<PostgresStorage<'a>>),
+    /// Generic implementation. Should be used for testing purposes only since it has performance penalty because
+    /// of the dynamic dispatch.
+    Boxed(Box<dyn ReadStorage + Send + 'a>),
```

Comment:
This variant and changes in the `ShadowStorage` are now technically unnecessary, but I'd argue they could still be useful (esp. `ShadowStorage`).